### PR TITLE
Generate versioned clientset

### LINKED ISSUE #20751: Generate client sets with versioned objects (instead of internal objects)

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix #20751

### CODE REVIEW

File: cmd/libs/go2idl/client-gen/generators/normalization/normalization.go [hunk 1]
```diff
@@ -29,3 +36,50 @@ func Version(version string) string {
...
+
+func GroupVersion(gv unversioned.GroupVersion) unversioned.GroupVersion {
+	return unversioned.GroupVersion{Group(gv.Group), Version(gv.Version)}
+}
+
+// Our assumption is that for the types.go that includes the internal version,
+// it's either located at.../pkg/api/ or .../pkg/apis/<group>/;
+// for the types.go that includes the versioned types, it's either located at
+// .../pkg/api/<version>/ or ../pkg/apis/<group>/<version>/.
+func ParseNormalizedGroupVersionFromPath(path string) (group, version string, err error) {
```
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/normalization/normalization.go]: I'll remove this function before merge. It probably is unnecessary.

File: pkg/client/typed/generated/extensions/v1beta1/thirdpartyresource.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+// A group's client should implement this interface.
+type ThirdPartyResourcesGetter interface {
+	ThirdPartyResources(namespace string) ThirdPartyResourceInterface
+}
+
+// ThirdPartyResourceInterface has methods to work with ThirdPartyResource resources.
+type ThirdPartyResourceInterface interface {
+	Create(*v1beta1.ThirdPartyResource) (*v1beta1.ThirdPartyResource, error)
+	Update(*v1beta1.ThirdPartyResource) (*v1beta1.ThirdPartyResource, error)
+	Delete(name string, options *api.DeleteOptions) error
```
Comment (caesarxuchao) [inline pkg/client/typed/generated/extensions/v1beta1/thirdpartyresource.go]: I'm not sure when we should use the versioned XXXOptions. I take a look at the manually written clients, they are using the internal XXXOptions.

**Comment by caesarxuchao** [comment]
@k8s-bot e2e test this. issue: #17874

File: cmd/libs/go2idl/client-gen/main.go [hunk 3]
```diff
@@ -48,24 +49,27 @@ func versionToPath(group string, version string) (path string) {
 	return
 }
 
-func parseInputVersions() ([]string, []unversioned.GroupVersion, error) {
+func parseInputVersions() ([]string, []unversioned.GroupVersion, map[unversioned.GroupVersion]string, error) {
```
Comment (lavalamp) [inline cmd/libs/go2idl/client-gen/main.go]: Please name the output parameters so this is self-documenting?

File: cmd/libs/go2idl/client-gen/generators/normalization/normalization.go [hunk 4]
```diff
@@ -29,3 +36,50 @@ func Version(version string) string {
 	}
 	return version
 }
+
+func GroupVersion(gv unversioned.GroupVersion) unversioned.GroupVersion {
+	return unversioned.GroupVersion{Group(gv.Group), Version(gv.Version)}
+}
+
+// Our assumption is that for the types.go that includes the internal version,
+// it's either located at.../pkg/api/ or .../pkg/apis/<group>/;
```
Comment (lavalamp) [inline cmd/libs/go2idl/client-gen/generators/normalization/normalization.go]: We will need to make this work with additional source directories. I.e., the `pkg/` part of those paths will need to be parameterizable. Eventually.

File: pkg/client/clientset_generated/release_1_2/clientset.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+limitations under the License.
+*/
+
+package release_1_2
+
+import (
+	"github.com/golang/glog"
+	core_v1 "k8s.io/kubernetes/pkg/client/typed/generated/core/v1"
+	extensions_v1beta1 "k8s.io/kubernetes/pkg/client/typed/generated/extensions/v1beta1"
+	unversioned "k8s.io/kubernetes/pkg/client/unversioned"
```
Comment (lavalamp) [inline pkg/client/clientset_generated/release_1_2/clientset.go]: Is this package necessary?

I see you use it for the config. Can the config be split out of the unversioned client? @krousey

**Comment by lavalamp** [comment]
Minor comments except about the client/unversioned package. I'll take it as is, but really the versioned clients should not import the unversioned client!

**Comment by caesarxuchao** [comment]
> I'll take it as is, but really the versioned clients should not import the unversioned client!

Right, I didn't notice the self-contradictory import. Apart from the unversioned.Config, we also need move several helper functions out of the unversioned package, like unversioned.DefaultKubernetesUserAgent().

**Comment by caesarxuchao** [comment]
Comments addressed, PR squashed. Self-applying the label per @lavalamp's comment.

File: cmd/libs/go2idl/client-gen/generators/client-generator.go [hunk 6]
```diff
@@ -38,6 +39,11 @@ type ClientGenArgs struct {
 	// unversioned package, which is part of our API. Tools like client-gen
 	// shouldn't depend on an API.
 	GroupVersions []unversioned.GroupVersion
+
+	// GropuVersionToInputPath is a map between GroupVersion and the path to
```
Comment (lavalamp) [inline cmd/libs/go2idl/client-gen/generators/client-generator.go]: s/Gropu/Group/

Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/client-generator.go]: Thanks. Fixed.

**Pull request merged**